clippy_lints/src/methods.rs

```diff
@@ -1105,6 +1125,50 @@ fn lint_iter_cloned_collect(cx: &LateContext, expr: &hir::Expr, iter_args: &[hir
     }
 }
 
+fn lint_fold_any(cx: &LateContext, expr: &hir::Expr, fold_args: &[hir::Expr]) {
+    // DONOTMERGE: What if this is just some other method called fold?
+    assert!(fold_args.len() == 3,
+        "Expected fold_args to have three entries - the receiver, the initial value and the closure");
+
+    if_chain! {
+        // Check if the initial value for the fold is the literal `false`
+        if let hir::ExprLit(ref lit) = fold_args[1].node;
+        if lit.node == ast::LitKind::Bool(false);
+
+        // Extract the body of the closure passed to fold
+        if let hir::ExprClosure(_, _, body_id, _, _) = fold_args[2].node;
+        let closure_body = cx.tcx.hir.body(body_id);
+        let closure_expr = remove_blocks(&closure_body.value);
+
+        // Extract the names of the two arguments to the closure
+        if let Some(first_arg_ident) = get_arg_name(&closure_body.arguments[0].pat);
+        if let Some(second_arg_ident) = get_arg_name(&closure_body.arguments[1].pat);
+
+        // Check if the closure body is of the form `acc || some_expr(x)`
+        if let hir::ExprBinary(ref bin_op, ref left_expr, ref right_expr) = closure_expr.node;
+        if bin_op.node == hir::BinOp_::BiOr;
+        if let hir::ExprPath(hir::QPath::Resolved(None, ref path)) = left_expr.node;
+        if path.segments.len() == 1 && &path.segments[0].name == &first_arg_ident;
+
+        then {
+            let right_source = snippet(cx, right_expr.span, "EXPR");
+
+            span_lint(
+                cx,
+                FOLD_ANY,
+                expr.span,
+                // TODO: don't suggest .any(|x| f(x)) if we can suggest .any(f)
+                &format!(
+                    ".fold(false, |{f}, {s}| {f} || {r})) is more succinctly expressed as .any(|{s}| {r})",
```

Comment:
Can you use `span_lint_and_sugg` here to get output like this:

```
error: This `.fold` can be more succinctly expressed as `.any`
   --> $DIR/methods.rs:390:13
    |
390 |     let _ = (0..3).fold(false, |acc, x| acc || x > 2);
    |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `.any(|x| x > 2)`
    |
    = note: `-D fold-any` implied by `-D warnings`
```

If you make sure the spans are right and exactly cover the `.fold(…)`, this link should work great with rustfix :)

Reply:
Sure, I'll just need to work out what it does first! (There's no doc comment at the moment)

Reply:
So, Rust's diagnostic messages (aka "fancy errors") allow you to add annotations to parts of the code. One special style of annotation is a "replacement suggestion". It looks just like the "help:" thing I posted above to a human, but [rustfix](https://github.com/killercup/rustfix) can parse it to automatically fix your code.

The mystical [`span_lint_and_sugg`](https://github.com/rust-lang-nursery/rust-clippy/blob/2f62d803aba284bfed6b5d5ef3a412496b9cc50c/clippy_lints/src/utils/mod.rs#L599) is a shortcut to create a new lint message and at the same time also add a suggestion. It's basically the same as `span_lint`, but you pass two parameters, `help: &str`, which is the "try: " in the example above, and `sugg: String`, which is the suggested replacement (something like `format!(".any(|{s}| {r})", …)` in this case here).

Reply:
Thanks anyway. I'll add a doc comment.

I've actually not quite got this working, as I've got a span pointing at the whole expression, including the range. What's the best way of restricting this to the `.fold(...)` part?